### PR TITLE
Introduce model hub

### PR DESCRIPTION
**Pull request by hendrik-obj**

### CODE REVIEW

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 1]
```diff
@@ -242,15 +246,168 @@ def ls(self):
...
+
+            if conversion_stack is None:
+                series = self._df.event_type == conversion_event
+            elif conversion_event is None:
+                series = conversion_stack.notnull()
+            else:
+                series = ((conversion_stack.notnull()) & (self._df.event_type == conversion_event))
+            return series.copy_override(name='conversion')
+
+        def conversions(self, name, partition='session_id'):
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:340]: I really think this and previous method need good docstrings before we ship. This is quite crucial, as we want to show how simple to use everything is. Personally, I find this method really hard to understand, but I tried to review anyway.

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 2]
```diff
@@ -242,15 +246,168 @@ def ls(self):
...
+            df = self._df.copy_override()
+            df['_conversion'] = df.mh.f.conversion(name)
+            df = df.materialize()
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:343]: this should not be required as `conversion()` only adds simple expressions?

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 3]
```diff
@@ -242,15 +246,168 @@ def ls(self):
...
+            exp = f"case when {{}} then row_number() over (partition by {{}}, {{}}) end"
+            df['_conversion_counter'] = df['_conversion'].copy_override(
+                dtype='int64',
+                expression=Expression.construct(exp, df['_conversion'], df[partition], df['_conversion']))
+            df = df.materialize()
+            exp = f"count({{}}) over (partition by {{}} order by {{}}, {{}})"
+            df = df.materialize()
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:350]: Duplicate, this one can go.

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 4]
```diff
@@ -242,15 +246,168 @@ def ls(self):
...
+        def conversions(self, name, partition='session_id'):
+            df = self._df.copy_override()
+            df['_conversion'] = df.mh.f.conversion(name)
+            df = df.materialize()
+            exp = f"case when {{}} then row_number() over (partition by {{}}, {{}}) end"
+            df['_conversion_counter'] = df['_conversion'].copy_override(
+                dtype='int64',
+                expression=Expression.construct(exp, df['_conversion'], df[partition], df['_conversion']))
+            df = df.materialize()
+            exp = f"count({{}}) over (partition by {{}} order by {{}}, {{}})"
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:349]: This one is supported by Bach directly:
`df['_conversion_counter'].count(df.groupby('_conversion_counter').sort_values([partition, 'moment']).window())` of the top of my head.

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 5]
```diff
@@ -325,36 +497,29 @@ def _prepare_sample(self, location_stack_column='location_stack'):
...
-    #     right_index=True)
+        sample_df = ObjectivFrame.get_instance(engine=df.engine,
+                                               base_node=new_base_node,
+                                               index_dtypes=df.index_dtypes,
+                                               dtypes=df.dtypes,
+                                               group_by=None,
+                                               order_by=None)
+
+        sample_df.time_aggregation = df.time_aggregation
+        sample_df.start_date = start_date
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:526]: where do these come from? For me this is broken.

File: analysis/bach_open_taxonomy/bach_open_taxonomy/stack/util.py [hunk 6]
```diff
@@ -7,9 +7,26 @@
 from sql_models.model import SqlModel
 
 
-def basic_feature_model(session_gap_seconds=1800) -> SqlModel:
+def _get_date_range(start_date, end_date):
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/stack/util.py:10]: Shouldn't we move this logic to `DateTimeOperation` aka `.dt`?

File: analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py [hunk 7]
```diff
@@ -242,15 +246,175 @@ def ls(self):
...
+        """
+        :param engine: a Sqlalchemy Engine for the database. If not given, env DSN is used to create one. If
+            that's not there, the default of 'postgresql://objectiv:@localhost:5432/objectiv' will be used.
+        :param time_aggregation: can be used to set a default aggregation timeframe interval that is used for
+            models that use aggregation. Ie. YYYY-MM-DD aggregates to days (dates). Setting it to None
+            aggregates over the entire selected dataset.
+        """
+        table_name = 'data'
+
+        if engine is None:
```
Comment (jansentom) [inline analysis/bach_open_taxonomy/bach_open_taxonomy/series/series_objectiv.py:413]: Added this to remove some complex looking boilerplate code.

File: analysis/notebooks/model-hub-demo-notebook.ipynb [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e97ff289",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from bach_open_taxonomy import ObjectivFrame"
```
Comment (jansentom) [inline analysis/notebooks/model-hub-demo-notebook.ipynb:21]: @borft @vincenthoogsteder all the sqlalchemy boilerplate here is gone. `from_table` below uses the DSN env variable to establish to connection itself if no engine is given.

**Review by jansentom — CHANGES_REQUESTED** [review]
Pretty cool stuff. I left some comments of which some really need attention, but should not be a lot of work.

File: analysis/notebooks/model-hub-demo-notebook.ipynb (hunk 8 above)
Comment (vincenthoogsteder) [reply analysis/notebooks/model-hub-demo-notebook.ipynb:21]: very nice

**Review by jansentom — APPROVED** [review]

**Pull request merged**